File: JAVA/Leetcode/CyclicallyRotateGrid.java
/*
Problem Statement:
You are given an m x n integer matrix grid, where m and n are both even integers, and an integer k.
A cyclic rotation of the matrix is done by cyclically rotating each layer in the matrix. 
To cyclically rotate a layer once, each element in the layer will take the place of the 
adjacent element in the counter-clockwise direction.
Return the matrix after applying k cyclic rotations to it.

Constraints:
m == grid.length
n == grid[i].length
2 <= m, n <= 50
Both m and n are even integers.
1 <= grid[i][j] <= 5000
1 <= k <= 10^9
*/

import java.util.Scanner;

public class CyclicallyRotateGrid {

    public static int[][] rotateGrid(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        
        // Since both m and n are even, the number of layers is half of the smaller dimension
        int numLayers = Math.min(m, n) / 2;
        
        for (int layer = 0; layer < numLayers; layer++) {
            // Define the boundaries for the current concentric layer
            int r1 = layer;
            int c1 = layer;
            int r2 = m - 1 - layer;
            int c2 = n - 1 - layer;
            
            // The perimeter size of the current layer
            int size = 2 * (r2 - r1) + 2 * (c2 - c1);
            int[] arr = new int[size];
            int idx = 0;
            
            // Extract the layer into a 1D array in Counter-Clockwise order:
            // 1. Left column (moving down)
            for (int i = r1; i < r2; i++) arr[idx++] = grid[i][c1];
            // 2. Bottom row (moving right)
            for (int j = c1; j < c2; j++) arr[idx++] = grid[r2][j];
            // 3. Right column (moving up)
            for (int i = r2; i > r1; i--) arr[idx++] = grid[i][c2];
            // 4. Top row (moving left)
            for (int j = c2; j > c1; j--) arr[idx++] = grid[r1][j];
            
            // Calculate effective rotations to avoid redundant cycles
            int rot = k % size;
            idx = 0;
            
            // Write the shifted elements back into the grid using the same CCW path
            // The new value at `idx` comes from `rot` steps behind it in the flattened array
            for (int i = r1; i < r2; i++) {
                grid[i][c1] = arr[(idx - rot + size) % size];
                idx++;
            }
            for (int j = c1; j < c2; j++) {
                grid[r2][j] = arr[(idx - rot + size) % size];
                idx++;
            }
            for (int i = r2; i > r1; i--) {
                grid[i][c2] = arr[(idx - rot + size) % size];
                idx++;
            }
            for (int j = c2; j > c1; j--) {
                grid[r1][j] = arr[(idx - rot + size) % size];
                idx++;
            }
        }
        
        return grid;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        try {
            System.out.println("Enter number of rows (m) and columns (n) - both must be EVEN:");
            String[] dims = scanner.nextLine().trim().split("\\s+");
            int m = Integer.parseInt(dims[0]);
            int n = Integer.parseInt(dims[1]);
            
            if (m % 2 != 0 || n % 2 != 0) {
                System.out.println("Both dimensions must be even integers.");
                return;
            }
            
            int[][] grid = new int[m][n];
            System.out.println("Enter the grid rows (numbers separated by spaces):");
            for (int i = 0; i < m; i++) {
                String[] rowStrs = scanner.nextLine().trim().split("\\s+");
                for (int j = 0; j < n; j++) {
                    grid[i][j] = Integer.parseInt(rowStrs[j]);
                }
            }
            
            System.out.println("Enter the number of rotations (k):");
            int k = Integer.parseInt(scanner.nextLine().trim());
            
            int[][] result = rotateGrid(grid, k);
            
            System.out.println("\nRotated Grid:");
            for (int i = 0; i < m; i++) {
                System.out.print("[");
                for (int j = 0; j < n; j++) {
                    System.out.print(result[i][j] + (j == n - 1 ? "" : ", "));
                }
                System.out.println("]");
            }
            
        } catch (Exception e) {
            System.out.println("Invalid input format detected.");
        } finally {
            scanner.close();
        }
    }
}

File: PYTHON/CyclicallyRotateGrid.py
'''
Problem Statement:
You are given an m x n integer matrix grid, where m and n are both even integers, and an integer k.
A cyclic rotation of the matrix is done by cyclically rotating each layer in the matrix. 
To cyclically rotate a layer once, each element in the layer will take the place of the 
adjacent element in the counter-clockwise direction.
Return the matrix after applying k cyclic rotations to it.

Constraints:
m == grid.length
n == grid[i].length
2 <= m, n <= 50
Both m and n are even integers.
1 <= grid[i][j] <= 5000
1 <= k <= 10^9
'''

def rotate_grid(grid: list[list[int]], k: int) -> list[list[int]]:
    m, n = len(grid), len(grid[0])
    num_layers = min(m, n) // 2
    
    for layer in range(num_layers):
        r1, c1 = layer, layer
        r2, c2 = m - 1 - layer, n - 1 - layer
        
        # Flatten the layer counter-clockwise
        arr = []
        
        # Left column (down)
        for i in range(r1, r2): arr.append(grid[i][c1])
        # Bottom row (right)
        for j in range(c1, c2): arr.append(grid[r2][j])
        # Right column (up)
        for i in range(r2, r1, -1): arr.append(grid[i][c2])
        # Top row (left)
        for j in range(c2, c1, -1): arr.append(grid[r1][j])
        
        size = len(arr)
        rot = k % size
        
        # Shift the 1D array effectively by combining slices
        # Shifting counter-clockwise pushes elements "forward" in the array,
        # which equates to a right-rotation in the 1D list structure.
        shifted_arr = arr[-rot:] + arr[:-rot]
        idx = 0
        
        # Write the shifted elements back in the exact same counter-clockwise path
        for i in range(r1, r2):
            grid[i][c1] = shifted_arr[idx]
            idx += 1
        for j in range(c1, c2):
            grid[r2][j] = shifted_arr[idx]
            idx += 1
        for i in range(r2, r1, -1):
            grid[i][c2] = shifted_arr[idx]
            idx += 1
        for j in range(c2, c1, -1):
            grid[r1][j] = shifted_arr[idx]
            idx += 1
            
    return grid

if __name__ == "__main__":
    try:
        dims_input = input("Enter number of rows (m) and columns (n) - both must be EVEN:\n").strip().split()
        m, n = int(dims_input[0]), int(dims_input[1])
        
        if m % 2 != 0 or n % 2 != 0:
            print("Both dimensions must be even integers.")
            exit()
            
        grid = []
        print("Enter the grid rows (numbers separated by spaces):")
        for _ in range(m):
            row = list(map(int, input().strip().split()))
            grid.append(row)
            
        k = int(input("Enter the number of rotations (k):\n").strip())
        
        result = rotate_grid(grid, k)
        
        print("\nRotated Grid:")
        for row in result:
            print(row)
            
    except ValueError:
        print("Invalid input detected. Please ensure all inputs are valid integers.")